examples/io/src/write_file.cc

```diff
@@ -0,0 +1,74 @@
+/* Copyright 2023 NVIDIA Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <fstream>
+
+#include "legate_library.h"
+#include "legateio.h"
+
+#include "core/utilities/dispatch.h"
+
+namespace legateio {
+
+namespace detail {
+
+struct write_fn {
+  template <legate::LegateTypeCode CODE>
+  void operator()(const legate::Store &input, const std::string &filename) {
+    using VAL = legate::legate_type_of<CODE>;
+
+    auto shape = input.shape<1>();
+    auto code = input.code<int64_t>();
+    size_t size = shape.volume();
+
+    // Store the type code and the number of elements in the array at the
+    // beginning of the file
+    std::ofstream out(filename,
+                      std::ios::binary | std::ios::out | std::ios::trunc);
+    out.write(reinterpret_cast<const char *>(&code), sizeof(int64_t));
+    out.write(reinterpret_cast<const char *>(&size), sizeof(size_t));
+
+    auto acc = input.read_accessor<VAL, 1>();
+    for (legate::PointInRectIterator it(shape); it.valid(); ++it) {
```

Comment:
This does not compile for me and needs an explicit <DIM>.  A better option might be a helper function that can properly do template argument deduction. Even better - fix this in Legion so that the shape can return an iter with a common begin/end idiom and this can be written with `for (auto point : shape.iter(false))`

Reply:
https://github.com/nv-legate/legate.core/pull/633/commits/964f05e88cfcd305a1db989b4c251a5a1f033028